src/bin/pg_autoctl/pgsql.c

```diff
@@ -117,6 +123,158 @@ pgsql_init(PGSQL *pgsql, char *url, ConnectionType connectionType)
 }
 
 
+/*
+ * pgsql_set_retry_policy sets the retry policy to the given maxT (maximum
+ * total time spent retrying), maxR (maximum number of retries, zero when not
+ * retrying at all, -1 for unbounded number of retries), and maxSleepTime to
+ * cap our exponential backoff with decorrelated jitter computation.
+ */
+void
+pgsql_set_retry_policy(PGSQL *pgsql,
+					   int maxT,
+					   int maxR,
+					   int maxSleepTime,
+					   int baseSleepTime)
+{
+	pgsql->retryPolicy.maxT = maxT;
+	pgsql->retryPolicy.maxR = maxR;
+	pgsql->retryPolicy.maxSleepTime = maxSleepTime;
+	pgsql->retryPolicy.baseSleepTime = baseSleepTime;
+
+	/* initialize a seed for our random number generator */
+	pg_srand48(time(0));
+
+	log_debug("pgsql_set_retry_policy(\"%s\"): maxT=%d maxR=%d cap=%d base=%d",
+			  pgsql->connectionString,
+			  pgsql->retryPolicy.maxT,
+			  pgsql->retryPolicy.maxR,
+			  pgsql->retryPolicy.maxSleepTime,
+			  pgsql->retryPolicy.baseSleepTime);
+}
+
+
+/*
+ * pgsql_set_default_retry_policy sets the default retry policy: no retry. We
+ * use the other default parameters but with a maxR of zero they don't get
+ * used.
+ *
+ * This is the retry policy that prevails in the main keeper loop.
+ */
+void
+pgsql_set_default_retry_policy(PGSQL *pgsql)
+{
+	(void) pgsql_set_retry_policy(pgsql,
+								  POSTGRES_PING_RETRY_TIMEOUT,
+								  0, /* do not retry by default */
+								  POSTGRES_PING_RETRY_CAP_SLEEP_TIME,
+								  POSTGRES_PING_RETRY_BASE_SLEEP_TIME);
+}
+
+
+/*
+ * pgsql_set_init_retry_policy sets the retry policy to 15 mins of total
+ * retrying time, unbounded number of attempts, and up to 2 seconds of sleep
+ * time in between attempts.
+ *
+ * This is the policy that we use in keeper_register_and_init. When using
+ * automated provisioning tools and frameworks, it might be that every node is
+ * provisionned concurrently and we might try to connect to the monitor before
+ * it's ready. In that case we want to retry for a long time.
+ */
+void
+pgsql_set_init_retry_policy(PGSQL *pgsql)
+{
+	(void) pgsql_set_retry_policy(pgsql,
+								  POSTGRES_PING_RETRY_TIMEOUT,
+								  -1, /* unbounded number of attempts */
+								  POSTGRES_PING_RETRY_CAP_SLEEP_TIME,
+								  POSTGRES_PING_RETRY_BASE_SLEEP_TIME);
+}
+
+
+/*
+ * pgsql_set_interactive_retry_policy sets the retry policy to 1 minute of
+ * total retrying time, unbounded number of attempts, and up to 2 seconds of
+ * sleep time in between attempts.
+ */
+void
+pgsql_set_interactive_retry_policy(PGSQL *pgsql)
```

Comment:
I think it's good to have this retry behaviour, but I think people should opt-in to it. For interactive usage fail-fast behaviour is usually what you want. So I think a flag like `--retry-connect` would be good to have to opt-in to it.

Also, do we want this retry behaviour for every interactive command or just a specific set? Currently it's not done for a command like:
```
pg_autoctl show state --pgdata node1
```

Reply:
So I have now made the interactive retry policy the default with a 2 seconds max time. Also, I made it so that users can set `PGCONNECT_TIMEOUT` in their environment to influence if they want to. We still manage that with our own retry strategy, though, and internally we use a timeout of 1 second only.

Reply:
Also all the commands are now using the interactive retry policy, only the main loop is having something different, and that's done in the main loop code itself.